File: examples/cpp/minimal-order-service/tests/test_usecases.cpp
#include <cassert>
#include <iostream>
#include <memory>
#include "application/usecases/place_order_use_case.hpp"
#include "application/ports/inbound/place_order_port.hpp"
#include "adapters/outbound/in_memory_repository.hpp"
#include "adapters/outbound/fake_payment_gateway.hpp"
#include "domain/order.hpp"

void test_place_order_use_case() {
    auto repository = std::make_shared<adapters::outbound::InMemoryOrderRepository>();
    auto paymentGateway = std::make_shared<adapters::outbound::FakePaymentGateway>();
    application::usecases::PlaceOrderUseCase useCase(repository, paymentGateway);
    
    application::ports::inbound::PlaceOrderCommand command{10000, "USD"};
    auto order = useCase.placeOrder(command);
    
    assert(order != nullptr);
    assert(order->status() == domain::OrderStatus::PAID);
    assert(order->total()->amount() == 10000);
    
    // Verify order was saved
    auto savedOrder = repository->findById(order->id());
    assert(savedOrder != nullptr);
    assert(savedOrder->id() == order->id());
    
    std::cout << "✓ test_place_order_use_case passed\n";
}

int main() {
    std::cout << "Running use case tests...\n";
    test_place_order_use_case();
    std::cout << "All use case tests passed!\n";
    return 0;
}

File: examples/python/minimal-order-service/tests/test_usecases.py
"""Tests for use cases."""
import pytest
from domain.order import OrderStatus
from application.usecases.place_order import PlaceOrderUseCase
from application.ports.inbound.place_order_port import PlaceOrderCommand
from adapters.outbound.in_memory_repository import InMemoryOrderRepository
from adapters.outbound.fake_payment_gateway import FakePaymentGateway


def test_place_order_use_case():
    """Test placing an order through the use case."""
    # Arrange
    repository = InMemoryOrderRepository()
    payment_gateway = FakePaymentGateway()
    use_case = PlaceOrderUseCase(repository, payment_gateway)
    
    # Act
    command = PlaceOrderCommand(amount=10000, currency="USD")
    order = use_case.place_order(command)
    
    # Assert
    assert order is not None
    assert order.status == OrderStatus.PAID
    assert order.total.amount == 10000
    assert order.total.currency == "USD"
    
    # Verify order was saved
    saved_order = repository.find_by_id(order.id)
    assert saved_order is not None
    assert saved_order.id == order.id
